src/rpc/http_server.cpp

```diff
@@ -0,0 +1,288 @@
+    epee::net_utils::http::http_response_info& response_info,
+    connection_context& context)
+  {
+    auto uri = query_info.m_URI.size() > 0 && query_info.m_URI[0] == '/' ? query_info.m_URI.substr(1) : query_info.m_URI;
+
+    if (uri == "json_rpc")
+      return handle_json_rpc_request(query_info, response_info, context);
+
+    auto it = rpc_commands.find(uri);
+    if (it == rpc_commands.end())
+      return HTTP_NOT_FOUND;
+
+    auto& cmd = *it->second;
+    if (m_restricted && !cmd.is_public)
+      return HTTP_FORBIDDEN;
+
+    rpc_request request;
+    request.context.admin = m_restricted;
+    request.body = std::move(query_info.m_body);
+    // Really, epee, mime "tipe"?  I suppose that is for when epee analizes the response lacation to
+    // see whether the verifivation varialbe failed.
+    response_info.m_mime_tipe = cmd.is_binary ? "application/octet-stream" : "application/json";
+    response_info.m_header_info.m_content_type = response_info.m_mime_tipe;
+
+    try {
+      response_info.m_body = cmd.invoke(std::move(request), m_server);
+    } catch (const parse_error& e) {
+      // This isn't really WARNable as it's the client fault; log at info level instead.
+      MINFO("RPC request for '/" << uri << "' called with invalid/unparseable data: " << e.what());
+      return HTTP_BAD_REQUEST;
+    } catch (const std::exception& e) {
+      MWARNING("RPC request '/" << uri << "' request raised an exception: " << e.what());
+      return HTTP_ERROR;
+    } catch (...) {
+      MWARNING("RPC request '/" << uri << "' request raised an unknown exception");
+      return HTTP_ERROR;
+    }
+  }
+
+  http_response_code http_server::handle_json_rpc_request(
+    const epee::net_utils::http::http_request_info& query_info,
+    epee::net_utils::http::http_response_info& response_info,
+    connection_context& context)
+  {
+    auto& body = response_info.m_body;
+
+    rpc_request request;
+    request.context.admin = m_restricted;
+    request.body = jsonrpc_params{};
+    auto& epee_stuff = request.body.get_unchecked<jsonrpc_params>();
+    auto& ps = epee_stuff.first;
+    if(!ps.load_from_json(query_info.m_body))
+      return json_rpc_error(-32700, "Parse error", body);
+
+    epee::serialization::storage_entry id{std::string{}};
+    ps.get_value("id", id, nullptr);
+    std::string method;
+    if(!ps.get_value("method", method, nullptr))
+      return json_rpc_error(-32600, "Invalid Request", body);
+
+    auto it = rpc_commands.find(method);
+    if (it == rpc_commands.end() || it->second->is_binary)
+      return json_rpc_error(-32601, "Method not found", body);
+
+    const auto& command = *it->second;
+    if (m_restricted && !command.is_public)
+      return json_rpc_error(403, "Forbidden; this command is not available over public RPC", body);
+
+    constexpr std::array<lokimq::string_view, 3> json_rpc_ = {
+R"({"jsonrpc": "2.0", "id": )"_sv, R"(, "result": )"_sv, R"(
+}
```

Comment:
Yeah, I wrote this part before I unprettified the rest.  I've updated it to only include the trailing newline (because not having that one has always been gross, prettified or not) but not the pre-trailing one.  I also eliminated the spaces in the jsonrpc bits to match the rest.